test/typescript/codeGeneration.js

```diff
@@ -0,0 +1,216 @@
+          hero: {
+            name: string,
+          } | null;
+        }
+      `);
+    });
+
+    it(`should generate simple nested query operations including input variables`, function() {
+      const context = this.compileFromSource(`
+        query HeroAndFriendsNames($episode: Episode) {
+          hero(episode: $episode) {
+            name
+            friends {
+              name
+            }
+          }
+        }
+      `);
+
+      const source = generateSource(context);
+
+      expect(source).to.include(stripIndent`
+        //  This file was automatically generated and should not be edited.
+
+        // The episodes in the Star Wars trilogy
+        export type Episode =
+          "NEWHOPE" | // Star Wars Episode IV: A New Hope, released in 1977.
+          "EMPIRE" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
+          "JEDI"; // Star Wars Episode VI: Return of the Jedi, released in 1983.
+
+
+        export interface HeroAndFriendsNamesQueryVariables {
+          episode: Episode | null;
+        }
+
+        export interface HeroAndFriendsNamesQuery {
+          hero: {
+            name: string,
+            friends: Array< {
+              name: string,
+            } > | null,
+          } | null;
+        }
+      `);
+    });
+
+    it(`should generate fragmented query operations`, function() {
+      const context = this.compileFromSource(`
+        query HeroAndFriendsNames {
+          hero {
+            name
+            ...HeroFriends
+          }
+        }
+
+        fragment HeroFriends on Character {
+          friends {
+            name
+          }
+        }
+      `);
+
+      const source = generateSource(context);
+
+      expect(source).to.include(stripIndent`
+        //  This file was automatically generated and should not be edited.
+
+        export interface HeroAndFriendsNamesQuery {
+          hero: HeroFriendsFragment & {
+            name: string,
+          } | null;
+        }
+
+        export interface HeroFriendsFragment {
+          friends: Array< {
+            name: string,
+          } > | null;
+        }
+      `);
+    });
+
+    it(`should generate query operations with inline fragments`, function() {
+      const context = this.compileFromSource(`
+        query HeroAndDetails {
+          hero {
+            name
+            ...HeroDetails
+          }
+        }
+
+        fragment HeroDetails on Character {
+          ... on Droid {
+            primaryFunction
+          }
+          ... on Human {
+            height
+          }
+        }
+      `);
+
+      const source = generateSource(context);
+
+      expect(source).to.include(stripIndent`
+        //  This file was automatically generated and should not be edited.
+
+        export interface HeroAndDetailsQuery {
+          hero: HeroDetailsFragment & {
+            name: string,
+          } | null;
+        }
+
+        export interface HeroDetailsFragment {
```

Comment:
Would it be more accurate to do something like:

```ts
{ primaryFunction: string | null } | { height: number | null }
```

So that it encodes that it's exactly one of the two combinations?